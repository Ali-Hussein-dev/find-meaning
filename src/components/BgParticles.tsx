import ParticleField from 'react-particles-webgl';
const config = {
  showCube: false,
  dimension: '2D',
  velocity: 1.2,
  boundaryType: 'passthru',
  antialias: false,
  direction: {
    xMin: -1,
    xMax: 1,
    yMin: -1,
    yMax: 1,
    zMin: -1,
    zMax: 1,
  },
  lines: {
    colorMode: 'rainbow',
    color: '#351CCB',
    transparency: 0.9,
    limitConnections: true,
    maxConnections: 20,
    minDistance: 110,
    visible: true,
  },
  particles: {
    colorMode: 'rainbow',
    color: '#3FB568',
    transparency: 0.9,
    shape: 'circle',
    boundingBox: 'canvas',
    count: 60,
    minSize: 20,
    maxSize: 50,
    visible: true,
  },
  cameraControls: {
    enabled: false,
    enableDamping: true,
    dampingFactor: 0.2,
    enableZoom: false,
    autoRotate: false,
    autoRotateSpeed: 0.3,
    resetCameraFlag: true,
  },
};
//=======================
export const BgParticles: React.FC = () => {
  // hooks
  //--------------------------------------
  // functions
  //--------------------------------------
  return <ParticleField config={config} />;
};
