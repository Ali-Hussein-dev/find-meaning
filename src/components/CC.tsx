import * as React from 'react';

interface Div extends React.ComponentPropsWithoutRef<'div'> {
  isTrue: any;
  fallback?: React.ReactElement;
}
//======================= CC stands for conditional component
export const CC: React.FC<Div> = ({
  isTrue,
  children,
  fallback = null,
  ...props
}) => {
  return isTrue ? <div {...props}>{children} </div> : fallback;
};
