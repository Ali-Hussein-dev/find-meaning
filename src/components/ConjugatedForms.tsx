import { Tooltip } from '@chakra-ui/react';
import * as React from 'react';
//=======================
type formLabels =
  | 'Infinitive'
  | 'Simple past'
  | 'Past participle'
  | 'Continous';
export type ConjugatedFormsT = [formLabels, string][];

//=======================
export const Pipe: React.FC = () => (
  <span className="text-blueGray-300">|</span>
);

export const ConjugatedForms: React.FC<{
  forms: ConjugatedFormsT;
}> = ({ forms }) => {
  //======================================return
  return (
    <>
      {forms.map((a, i) => (
        <React.Fragment key={i}>
          <Tooltip hasArrow label={[a[0]]} aria-label="a tooltip">
            {a[1]}
          </Tooltip>
          {forms.length - 1 > i ? <Pipe /> : null}
        </React.Fragment>
      ))}
    </>
  );
};
