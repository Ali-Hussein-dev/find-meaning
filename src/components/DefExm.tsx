//=======================
export const DefExm: React.FC<{
  def: string;
  exm: string | string[];
  context?: string[];
}> = ({ def, exm, context }) => {
  return (
    <div className="w-11/12 mb-3 text-blueGray-600">
      <p>
        <span className=" text-lightBlue-500">{context?.join()} </span>
        {def}
      </p>
      <>
        {typeof exm === 'string' ? (
          <p className="ml-1 italic font-light text-trueGray-600">{exm}</p>
        ) : Array.isArray(exm) ? (
          exm?.map((str, i) => (
            <p className="ml-1 italic font-light text-trueGray-600" key={i}>
              {str}
            </p>
          ))
        ) : null}
      </>
    </div>
  );
};
