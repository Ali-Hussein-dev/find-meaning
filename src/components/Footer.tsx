export const Footer: React.FC = () => {
  //======================================return
  return (
    <div className="w-full h-12 text-lg font-light center bg-blueGray-100 gap-x-4">
      &copy; {new Date().getFullYear()} <span>|</span> Find-Meaning
    </div>
  );
};
