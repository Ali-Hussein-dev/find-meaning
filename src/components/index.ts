export * from './SearchBar';
export * from './Pronunciation';
export * from './WikContainer';
export * from './WikiLexeme';
export * from './Gifs';
export * from './CC';
export * from './Input';
export * from './DefExm';
export * from './UrbanContainer';
export * from './BgParticles';
export * from './leaf-components/AsyncStatus';
export * from './leaf-components/HighlightSubstring';
export * from './leaf-components/FeedbackSubmitted';
export * from './FeedbackDrawer';
export * from './Footer';
export * from './leaf-components/CustomSkeleton';
