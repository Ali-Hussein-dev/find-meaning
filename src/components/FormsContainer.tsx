import { CondComp } from '@/components/index';

export const FormsContainer: React.FC<{ baseCond: boolean }> = ({
  baseCond,
  children,
}) => (
  <CondComp
    baseCond={baseCond}
    className="flex flex-wrap items-start font-semibold rounded gap-x-2 gap-y-1 text-blueGray-600"
  >
    <span className="label">forms</span>
    {children}
  </CondComp>
);
