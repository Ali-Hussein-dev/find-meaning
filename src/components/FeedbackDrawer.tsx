import * as React from 'react';
import { nanoid } from 'nanoid';
import store from 'store2';
import { isMobile } from 'react-device-detect';
import { fetcherPost } from 'src/utils';
import { useQuery } from 'react-query';
import { CondComp, FeedbackSubmitted } from '@/components/index';
import { MdClear, MdSend } from 'react-icons/md';
import {
  Input,
  Textarea,
  Stack,
  useDisclosure,
  Spinner,
  Drawer,
  DrawerBody,
  DrawerFooter,
  DrawerHeader,
  DrawerOverlay,
  DrawerContent,
} from '@chakra-ui/react';

//--------------------------------------setIdentiefer
(function () {
  if (!store.get('identiefer')) {
    store.set('identiefer', { _id: nanoid(), feedbackCount: 0 });
  }
  return;
})();
//--------------------------------------updateIdentieferProperty
const updateIdentieferProperty = (
  identieferObj: {
    _id: string;
    feedbackCount: number;
  },
  keyValue,
) => {
  store.set('identiefer', { ...identieferObj, ...keyValue });
};
const feedbackCount = store.get('identiefer').feedbackCount;
//=======================FeedbackSubmitted

//=======================FeedbackDrawer
export const FeedbackDrawer: React.FC = () => {
  //--------------------------------------hooks
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [nameValue, setNameValue] = React.useState('');
  const [emailValue, setEmailValue] = React.useState('');
  const [textareaValue, setTextareaValue] = React.useState('');
  const [enabledFetching, setEnabledFetching] = React.useState(false);
  const { status } = useQuery(
    'feedback',
    () =>
      fetcherPost('api/feedback', {
        _id: store.get('identiefer')._id,
        feedback: {
          userName: nameValue,
          email: emailValue,
          feedback: textareaValue,
          createdAt: new Date(),
          stars: undefined,
        },
      }),
    {
      enabled: enabledFetching,
      staleTime: Infinity,
      onSuccess: () => {
        updateIdentieferProperty(store.get('identiefer'), {
          feedbackCount: store.get('identiefer').feedbackCount + 1,
        });
      },
    },
  );
  //--------------------------------------functions
  const handleSubmit = async (e) => {
    e.preventDefault();
    setEnabledFetching(true);
  };
  //======================================return
  return (
    feedbackCount < 3 && (
      <div className="w-full mb-3 text-center">
        <Drawer
          isOpen={isOpen}
          placement={isMobile ? 'bottom' : 'right'}
          onClose={onClose}
        >
          <DrawerOverlay>
            <DrawerContent data-testid="feedback-container" bg={''}>
              <DrawerHeader className="bg-lightBlue-800">
                <div className="">
                  <h2 className="text-base text-lightBlue-300">Feedback</h2>
                  <h2 className="text-xl font-normal text-blueGray-50">
                    Your feedback is valuable to us
                  </h2>
                </div>
              </DrawerHeader>
              <DrawerBody className="bg-blueGray-50 text-blueGray-700">
                <form onSubmit={handleSubmit}>
                  <CondComp
                    baseCond={status !== 'success'}
                    fallback={
                      <FeedbackSubmitted msg={'Thank you for your feedback!'} />
                    }
                    className="py-3 "
                  >
                    <Stack spacing={4}>
                      <label htmlFor="name">
                        <Input
                          isRequired
                          id="name"
                          variant="flushed"
                          value={nameValue}
                          onChange={(e) => setNameValue(e.target.value)}
                          type="text"
                          placeholder="name*"
                        />
                      </label>
                      <label htmlFor="email">
                        <Input
                          id="email"
                          isRequired
                          variant="flushed"
                          type="email"
                          placeholder="email*"
                          value={emailValue}
                          onChange={(e) => setEmailValue(e.target.value)}
                        />
                      </label>
                      <label htmlFor="feedback-text">
                        <Textarea
                          id="feedback-text"
                          variant="flushed"
                          isRequired
                          value={textareaValue}
                          onChange={(e) => setTextareaValue(e.target.value)}
                          placeholder="e.g. ask for a nice feature*"
                        />
                      </label>
                    </Stack>
                  </CondComp>
                </form>
              </DrawerBody>
              <DrawerFooter className="flex justify-between w-full gap-x-1 bg-blueGray-50">
                <button
                  type="button"
                  onClick={onClose}
                  className="btn flex-grow h-8 rounded-lg press-effect focus:ring-2 center text-blueGray-600 ring-yellow-600"
                >
                  <span className="center gap-x-2">
                    <MdClear size="20" />
                    Cancel
                  </span>
                </button>
                {status !== 'success' && (
                  <button
                    onClick={handleSubmit}
                    type="submit"
                    name="submit-feedback"
                    data-testid="submit-feedback"
                    className="btn flex-grow h-8 text-white rounded-lg bg-lightBlue-600 disabled:bg-blueGray-200 disabled:text-blueGray-400 disabled:cursor-not-allowed focus:ring-2 center gap-x-1 ring-yellow-600"
                    disabled={
                      !(!!emailValue && !!textareaValue && !!nameValue) ||
                      status === 'loading'
                    }
                  >
                    {status === 'loading' ? (
                      <Spinner size="sm" />
                    ) : (
                      <span className="center gap-x-2">
                        <MdSend size="15" />
                        Send
                      </span>
                    )}
                  </button>
                )}
              </DrawerFooter>
            </DrawerContent>
          </DrawerOverlay>
        </Drawer>
        <button
          type="button"
          onClick={onOpen}
          className="p-1 font-semibold underline rounded focus:outline-none text-blueGray-300 focus:ring-2 ring-lightBlue-500 hover:text-lightBlue-500"
        >
          Was this helpful?
        </button>
      </div>
    )
  );
};
