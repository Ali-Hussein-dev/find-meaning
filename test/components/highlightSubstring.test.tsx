import { HighlightSubstring } from '@/components/index';
import { render, screen } from '../testUtils';
describe('Highlight unmatched characters when typing', () => {
  it('Should unbold inputValue characters', () => {
    render(
      <HighlightSubstring inputValue={'sugg'} suggestion={'suggestion'} />,
    );
    const matched = screen.getByText('sugg');
    expect(matched).toHaveAttribute('class', 'text-blueGray-300');
  });
  //--------------------------------------
  it('Should display unhighlighted characters WHEN inputValue is empty', () => {
    const { container } = render(
      <HighlightSubstring inputValue="" suggestion="suggestion" />,
    );
    expect(container).toMatchSnapshot();
    expect(container).toMatchInlineSnapshot(`
      <div>
        <span>
          <span
            class="text-blueGray-300"
          />
          <span>
            suggestion
          </span>
        </span>
      </div>
    `);
  });
  //--------------------------------------
});
