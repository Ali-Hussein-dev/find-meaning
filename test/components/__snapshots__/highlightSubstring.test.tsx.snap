// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight unmatched characters when typing Should display unhighlighted characters WHEN inputValue is empty 1`] = `
<div>
  <span>
    <span
      class="text-blueGray-300"
    />
    <span>
      suggestion
    </span>
  </span>
</div>
`;
